### PR TITLE
DR-1125 Remove child-process.execSync()

### PR DESCRIPTION
**Pull request by emab1**
## Description

Added small code smell fixes and removed child-process.execSync() calls from UI tests.

## Context

execSync() is security sensitive and creates a hotspot in Sonar Cloud. Sonar Cloud shows no security hotspots with this PR:
Screenshot 2024-10-08 at 14 27 46

## Type of changes

- [ ] Refactoring (non-breaking change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would change existing functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)

## Checklist

- [ ] I am familiar with the [contributing guidelines](../docs/CONTRIBUTING.md)
- [ ] I have followed the code style of the project
- [ ] I have added tests to cover my changes
- [ ] I have updated the documentation accordingly
- [ ] This PR is a result of pair or mob programming

---

## Sensitive Information Declaration

To ensure the utmost confidentiality and protect your and others privacy, we kindly ask you to NOT including [PII (Personal Identifiable Information) / PID (Personal Identifiable Data)](https://digital.nhs.uk/data-and-information/keeping-data-safe-and-benefitting-the-public) or any other sensitive data in this PR (Pull Request) and the codebase changes. We will remove any PR that do contain any sensitive information. We really appreciate your cooperation in this matter.

- [ ] I confirm that neither PII/PID nor sensitive data are included in this PR and the codebase changes.

### CODE REVIEW

File: tests/ui/src/utilities/s3Helper.ts [hunk 1]
```diff
@@ -1,28 +1,64 @@
...
+  try {
+    const fileContent = await readFile(sourceFilePath);
+    const command = new PutObjectCommand({
+      Bucket: bucket,
+      Key: targetFile,
+      Body: fileContent,
+    });
+
+    const response = await client.send(command);
+    console.log('Success', response);
```
Comment (AITR1) [inline tests/ui/src/utilities/s3Helper.ts:23]: For standard development having these logs might be usful but I don't think they are really necessary here as they are just utils for tests

File: tests/ui/src/utilities/s3Helper.ts [hunk 2]
```diff
@@ -1,28 +1,64 @@
...
+import { pipeline, Readable } from 'stream';
+import { promisify } from 'util';
 
 const bucket = `nhse-uec-cm-ui-test-bucket-${process.env.ENV}${process.env.WORKSPACE}`;
 
-export function addObject(sourceFilePath: string, targetFile: string) {
-  return require('child_process').execSync(
-    `aws s3api put-object --bucket ${bucket} --key ${targetFile} --body ${sourceFilePath}`
-)};
+const pipelineAsync = promisify(pipeline);
```
Comment (AITR1) [inline tests/ui/src/utilities/s3Helper.ts:9]: Can this be moved to where its being used?

File: tests/ui/src/utilities/globalSetup.ts [hunk 3]
```diff
@@ -13,9 +13,16 @@ async function globalSetup(config: FullConfig) {
...
+    await getCloudFrontUrl(region, env, workspace).then((distribution) =>{
+      if (distribution) {
+        const baseUrl = distribution.DomainName;
+        process.env.baseUrl = baseUrl;
+      } else {
+        console.log('No domain name found');
+      }
+    }).catch((error) => {
+      console.error('Error getting cloud front url', error);
+    });
```
Comment (Tomdango) [inline tests/ui/src/utilities/globalSetup.ts:25]: You're doing all your error handling in the function above, so should be able to simply this with just

```ts
const distribution = await getCloudFrontUrl(region, env, workspace);
if (distribution) {
    process.env.baseUrl = distribution.DomainName;
} else {
    console.log('No domain name found');
}
```

File: tests/ui/src/utilities/s3Helper.ts [hunk 4]
```diff
@@ -1,28 +1,60 @@
...
+      Bucket: bucket,
+      Key: targetFile,
+      Body: fileContent,
+    });
+
+    await client.send(command);
+  } catch (error) {
+    console.error('Error', error);
+  }
+};
```
Comment (Tomdango) [inline tests/ui/src/utilities/s3Helper.ts:24]: The original `addObject` function would crash out if it got a non-zero exit code, so I'd just remove the wrapping try-catch block and let this error bubble up and fail the test

**Review by mmg-nhse — DISMISSED** [review]

File: tests/ui/src/utilities/cloudFront.ts [hunk 5]
```diff
@@ -1,5 +1,32 @@
...
+          DomainName: filteredDistribution.DomainName,
+        };
+      } else {
+        console.log('No matching distribution found');
+        return null;
+      }
+    } else {
+      console.log('No distributions found');
+      return null;
+    }
```
Comment (Tomdango) [inline tests/ui/src/utilities/cloudFront.ts:27]: Nitpick/Suggestion - using negations instead can help make your code clearer and less nested. Also - if we just need the domain name, I'd return that directly.

 For example:
```suggestion
    const distributions = response.DistributionList?.Items;
    
    if (!distributions) {
        console.log('No distributions found');
        return null;
    }
    
    const filteredDistribution = distributions.find(dist =>
        dist.Origins?.Items?.[0].DomainName === `nhse-uec-cm-${env}-front-end${workspace}.s3.${region}.amazonaws.com`
    );
    
    if (!filteredDistribution) {
        console.log('No matching distribution found');
        return null;
    }
    
    return filteredDistribution.DomainName;
```

File: tests/ui/src/utilities/globalSetup.ts [hunk 6]
```diff
@@ -12,9 +12,13 @@ async function globalSetup() {
...
-    const distribution = getCloudFrontUrl(region, env, workspace);
-    const baseUrl = JSON.parse(distribution).DomainName;
-    process.env.baseUrl = baseUrl;
+    const distribution = await getCloudFrontUrl(region, env, workspace);
+    if (distribution) {
+      const baseUrl = distribution.DomainName;
+      process.env.baseUrl = baseUrl;
+    } else {
+      console.log('No domain name found');
+    }
```
Comment (Tomdango) [inline tests/ui/src/utilities/globalSetup.ts:21]: If you directly return the domain name, you can simplify this a good bit.

```suggestion
    process.env.baseUrl = await getCloudFrontUrl(region, env, workspace);
    if (!process.env.baseUrl) {
      console.log('No domain name found');
    }
```

**Review by Tomdango — DISMISSED** [review]
Changes look good - approved. Just a couple of coding style comments - no hard requirement to implement.

**Review by mmg-nhse — APPROVED** [review]

File: tests/ui/src/utilities/globalSetup.ts [hunk 7]
```diff
@@ -14,8 +14,7 @@ async function globalSetup() {
     const region = process.env.REGION as string;
     const distribution = await getCloudFrontUrl(region, env, workspace);
     if (distribution) {
-      const baseUrl = distribution.DomainName;
-      process.env.baseUrl = baseUrl;
+      process.env.baseUrl = distribution;
```
Comment (AITR1) [inline tests/ui/src/utilities/globalSetup.ts:17]: It was
```
 const baseUrl = distribution.DomainName;
 process.env.baseUrl = baseUrl;
```
I think we are losing the .DomainNam in "process.env.baseUrl = distribution;"?
Should it be " process.env.baseUrl = distribution.DomainName;"?

Comment (AITR1) [reply tests/ui/src/utilities/globalSetup.ts:17]: Actually I think I misread the change in tests/ui/src/utilities/cloudFront.ts

**Review by AITR1 — APPROVED** [review]

**Comment by AITR1** [comment]
LGTM, approving in lieu of Tomdango

**Review by mmg-nhse — APPROVED** [review]

**Pull request merged**